## Data Wrangling:

In this section,I begin with the data csv file, which we obtained from the US Census Bureau. As a recall, the data 
contains mostly numerical variables and few categorical. The major objective of this section is to understand the 
general trends in the data, and gain some quick insights and relation between variables that will help in machine 
learning later for the prediction of california house price. The statistical significance of these observations will be tested in the next section Exploratory Data 
Analysis. 

In [2]:
import warnings
warnings.filterwarnings("ignore")
import time
import random
from math import *
import operator
import pandas as pd
import numpy as np
from scipy import stats

# import plotting libraries
import matplotlib
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
%matplotlib inline 

import seaborn as sns
sns.set(style="white", color_codes=True)
sns.set(font_scale=1.5)

# load make_blobs to simulate data
from sklearn.datasets import make_blobs
from sklearn.datasets import make_classification


# pre-processing:
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import scale
from sklearn.preprocessing import minmax_scale
from sklearn.preprocessing import Normalizer
from sklearn.preprocessing.data import QuantileTransformer
from sklearn.preprocessing import Imputer
from IPython.display import Markdown, display
from sklearn.preprocessing import LabelEncoder
import os
import statistics


In [29]:
df_Calidata=pd.read_csv("C://Users//disoj//Desktop//house_price.csv")

In [30]:
# number of rows and columns:
df_Calidata.shape

(20640, 10)

In [31]:
# Top 5 rows:
df_Calidata.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                  41          880           129.0   
1    -122.22     37.86                  21         7099          1106.0   
2    -122.24     37.85                  52         1467           190.0   
3    -122.25     37.85                  52         1274           235.0   
4    -122.25     37.85                  52         1627           280.0   

   population  households  median_income ocean_proximity  median_house_value  
0         322         126         8.3252        NEAR BAY              452600  
1        2401        1138         8.3014        NEAR BAY              358500  
2         496         177         7.2574        NEAR BAY              352100  
3         558         219         5.6431        NEAR BAY              341300  
4         565         259         3.8462        NEAR BAY              342200

In [32]:
# summary statistics of numeric columns:
df_Calidata.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [33]:
df_Calidata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
longitude             20640 non-null float64
latitude              20640 non-null float64
housing_median_age    20640 non-null int64
total_rooms           20640 non-null int64
total_bedrooms        20433 non-null float64
population            20640 non-null int64
households            20640 non-null int64
median_income         20640 non-null float64
ocean_proximity       20640 non-null object
median_house_value    20640 non-null int64
dtypes: float64(4), int64(5), object(1)
memory usage: 1.6+ MB


In [34]:
df_Calidata['ocean_proximity'].value_counts()

<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: ocean_proximity, dtype: int64

'ocean_proximity' is a categorical variable, which has 5 branches. 'Island' branch of 'ocean_proximity' has only 5 number of records in a dataset.

In [35]:
# Checking missing values
df_Calidata.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
ocean_proximity         0
median_house_value      0
dtype: int64

Total bedrooms has 207 missing values. Mean doesn't makes sense for total bedrooms, so replacing missing values with most frequenty occured number, i.e. mode of the numerical variables of total bedrooms.

In [36]:
mode=statistics.mode(df_Calidata['total_bedrooms'])

In [37]:
mode

280.0

In [38]:
# Filling missing 'total_bedrooms' values with mode since mean doesn't make sense for total bedrooms:
df_Calidata['total_bedrooms']=df_Calidata['total_bedrooms'].fillna(df_Calidata['total_bedrooms'].mode()[0])

In [39]:
# Again Checking missing values
df_Calidata.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
ocean_proximity       0
median_house_value    0
dtype: int64

In [40]:
df_Calidata.dtypes

longitude             float64
latitude              float64
housing_median_age      int64
total_rooms             int64
total_bedrooms        float64
population              int64
households              int64
median_income         float64
ocean_proximity        object
median_house_value      int64
dtype: object

In [41]:
df_Calidata['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [42]:
df_Calidata["ocean_proximity"] = df_Calidata["ocean_proximity"].astype('category')

df_Calidata.dtypes

longitude              float64
latitude               float64
housing_median_age       int64
total_rooms              int64
total_bedrooms         float64
population               int64
households               int64
median_income          float64
ocean_proximity       category
median_house_value       int64
dtype: object


### Creating new features:
#### rooms per household 
#### bedrooms per room 
#### population per household 
We can check if these additional features have any additional impact on the median house price.

In [43]:
df_Calidata['rooms_per_household']=df_Calidata['total_rooms']/df_Calidata['households']
df_Calidata['bedrooms_per_room']=df_Calidata['total_bedrooms']/df_Calidata['total_rooms']
df_Calidata['population_per_household']=df_Calidata['population']/df_Calidata['households']

In [44]:
df_Calidata.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                  41          880           129.0   
1    -122.22     37.86                  21         7099          1106.0   
2    -122.24     37.85                  52         1467           190.0   
3    -122.25     37.85                  52         1274           235.0   
4    -122.25     37.85                  52         1627           280.0   

   population  households  median_income ocean_proximity  median_house_value  \
0         322         126         8.3252        NEAR BAY              452600   
1        2401        1138         8.3014        NEAR BAY              358500   
2         496         177         7.2574        NEAR BAY              352100   
3         558         219         5.6431        NEAR BAY              341300   
4         565         259         3.8462        NEAR BAY              342200   

   rooms_per_household  bedrooms_per_room  population_per_household  
0             6.984127           0.146591                  2.555556  
1             6.238137           0.155797                  2.109842  
2             8.288136           0.129516                  2.802260  
3             5.817352           0.184458                  2.547945  
4             6.281853           0.172096                  2.181467

In [45]:
# Lets get dummies data for Categorical variables.
df_Calidata1 = pd.get_dummies(df_Calidata, prefix_sep='-',drop_first=True,columns=["ocean_proximity"])

In [46]:
df_Calidata1.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                  41          880           129.0   
1    -122.22     37.86                  21         7099          1106.0   
2    -122.24     37.85                  52         1467           190.0   
3    -122.25     37.85                  52         1274           235.0   
4    -122.25     37.85                  52         1627           280.0   

   population  households  median_income  median_house_value  \
0         322         126         8.3252              452600   
1        2401        1138         8.3014              358500   
2         496         177         7.2574              352100   
3         558         219         5.6431              341300   
4         565         259         3.8462              342200   

   rooms_per_household  bedrooms_per_room  population_per_household  \
0             6.984127           0.146591                  2.555556   
1             6.238137           0.155797                  2.109842   
2             8.288136           0.129516                  2.802260   
3             5.817352           0.184458                  2.547945   
4             6.281853           0.172096                  2.181467   

   ocean_proximity-INLAND  ocean_proximity-ISLAND  ocean_proximity-NEAR BAY  \
0                       0                       0                         1   
1                       0                       0                         1   
2                       0                       0                         1   
3                       0                       0                         1   
4                       0                       0                         1   

   ocean_proximity-NEAR OCEAN  
0                           0  
1                           0  
2                           0  
3                           0  
4                           0

In [47]:
df_Calidata1.dtypes

longitude                     float64
latitude                      float64
housing_median_age              int64
total_rooms                     int64
total_bedrooms                float64
population                      int64
households                      int64
median_income                 float64
median_house_value              int64
rooms_per_household           float64
bedrooms_per_room             float64
population_per_household      float64
ocean_proximity-INLAND          uint8
ocean_proximity-ISLAND          uint8
ocean_proximity-NEAR BAY        uint8
ocean_proximity-NEAR OCEAN      uint8
dtype: object

In [48]:
df_Calidata1.shape

(20640, 16)

Lets apply standard Scaler in this case: This Standardizes features by removing the mean and scaling to unit variance.
Centering and scaling happen independently on each feature by computing the relevant statistics on the samples in the 
training set. Mean and standard deviation are then stored to be used on later data using the transform method.


In [49]:
scaler=StandardScaler(copy=True,with_mean=True,with_std=True
                     ).fit(df_Calidata1)
rescaled_dfnhouse=scaler.transform(df_Calidata1)

In [50]:
colnames = ['longitude','latitude','housing_median_age','total_rooms','total_bedrooms','population','households','median_income','median_house_price','ocean_prox_INLAND','ocean_prox_ISLAND','ocean_prox_NEAR BAY','ocean_prox_NEAR OCEAN','rooms_per_household','bedrooms_per_room','population_per_household']
df_Calidata2= pd.DataFrame(rescaled_dfnhouse, columns=colnames)

In [51]:
df_Calidata2.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0  -1.327835  1.052548            0.982143    -0.804819       -0.967245   
1  -1.322844  1.043185           -0.607019     2.045890        1.358707   
2  -1.332827  1.038503            1.856182    -0.535746       -0.822021   
3  -1.337818  1.038503            1.856182    -0.624215       -0.714889   
4  -1.337818  1.038503            1.856182    -0.462404       -0.607758   

   population  households  median_income  median_house_price  \
0   -0.974429   -0.977033       2.344766            2.129631   
1    0.861439    1.669961       2.332238            1.314156   
2   -0.820777   -0.843637       1.782699            1.258693   
3   -0.766028   -0.733781       0.932968            1.165100   
4   -0.759847   -0.629157      -0.012881            1.172900   

   ocean_prox_INLAND  ocean_prox_ISLAND  ocean_prox_NEAR BAY  \
0           0.628559          -1.088291            -0.049597   
1           0.327041          -0.936892            -0.092512   
2           1.155620          -1.369108            -0.025843   
3           0.156966          -0.465514            -0.050329   
4           0.344711          -0.668831            -0.085616   

   ocean_prox_NEAR OCEAN  rooms_per_household  bedrooms_per_room  \
0              -0.681889            -0.015566           2.830742   
1              -0.681889            -0.015566           2.830742   
2              -0.681889            -0.015566           2.830742   
3              -0.681889            -0.015566           2.830742   
4              -0.681889            -0.015566           2.830742   

   population_per_household  
0                 -0.384466  
1                 -0.384466  
2                 -0.384466  
3                 -0.384466  
4                 -0.384466

In [52]:
# extract the data, into numpy array:
X_features =df_Calidata2.drop(['median_house_price'],axis=1)
y_actual=df_Calidata2['median_house_price']

In [53]:
X_features.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0  -1.327835  1.052548            0.982143    -0.804819       -0.967245   
1  -1.322844  1.043185           -0.607019     2.045890        1.358707   
2  -1.332827  1.038503            1.856182    -0.535746       -0.822021   
3  -1.337818  1.038503            1.856182    -0.624215       -0.714889   
4  -1.337818  1.038503            1.856182    -0.462404       -0.607758   

   population  households  median_income  ocean_prox_INLAND  \
0   -0.974429   -0.977033       2.344766           0.628559   
1    0.861439    1.669961       2.332238           0.327041   
2   -0.820777   -0.843637       1.782699           1.155620   
3   -0.766028   -0.733781       0.932968           0.156966   
4   -0.759847   -0.629157      -0.012881           0.344711   

   ocean_prox_ISLAND  ocean_prox_NEAR BAY  ocean_prox_NEAR OCEAN  \
0          -1.088291            -0.049597              -0.681889   
1          -0.936892            -0.092512              -0.681889   
2          -1.369108            -0.025843              -0.681889   
3          -0.465514            -0.050329              -0.681889   
4          -0.668831            -0.085616              -0.681889   

   rooms_per_household  bedrooms_per_room  population_per_household  
0            -0.015566           2.830742                 -0.384466  
1            -0.015566           2.830742                 -0.384466  
2            -0.015566           2.830742                 -0.384466  
3            -0.015566           2.830742                 -0.384466  
4            -0.015566           2.830742                 -0.384466

Correlations between features and target:

In [54]:
from scipy.stats import pearsonr
features=['longitude','latitude','housing_median_age','total_rooms','total_bedrooms','population','households','median_income','ocean_prox_INLAND','ocean_prox_ISLAND','ocean_prox_NEAR BAY','ocean_prox_NEAR OCEAN','rooms_per_household','bedrooms_per_room','population_per_household']
target=df_Calidata2['median_house_price'].name
correlations = {}
for f in features:
    data_temp = df_Calidata2[[f,target]]
    x1 = data_temp[f].values
    x2 = data_temp[target].values
    key = f + ' vs ' + target
    correlations[key] = pearsonr(x1,x2)[0]
    

In [55]:
data_correlations = pd.DataFrame(correlations, index=['Value']).T
data_correlations.loc[data_correlations['Value'].abs().sort_values(ascending=False).index]

Value
median_income vs median_house_price             0.688075
ocean_prox_NEAR OCEAN vs median_house_price    -0.484859
ocean_prox_ISLAND vs median_house_price        -0.247152
bedrooms_per_room vs median_house_price         0.160284
ocean_prox_INLAND vs median_house_price         0.151948
latitude vs median_house_price                 -0.144160
population_per_household vs median_house_price  0.141862
total_rooms vs median_house_price               0.134153
housing_median_age vs median_house_price        0.105623
households vs median_house_price                0.065843
total_bedrooms vs median_house_price            0.049406
longitude vs median_house_price                -0.045967
population vs median_house_price               -0.024650
ocean_prox_NEAR BAY vs median_house_price      -0.023737
rooms_per_household vs median_house_price       0.023416

The above correlations indicates that population_per_household and bedrooms_per_room have better correlation with median_house_price than total population and total_bedrooms.Best regressor for the median house price is median_income in this case.

#### Feature Selection Processes based on Multicollenearity check:

In [56]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
# For each X_features, calculate VIF and save in dataframe
VIF = pd.DataFrame()

VIF["VIF_Factor"] = [variance_inflation_factor(X_features.values, i) for i in range(X_features.shape[1])]

VIF["feature"] = X_features.columns

In [57]:
VIF.sort_values(['VIF_Factor'], ascending=False).round(5)

VIF_Factor                   feature
4     33.65446            total_bedrooms
6     32.73443                households
1     20.42827                  latitude
0     18.39216                 longitude
3     17.14962               total_rooms
5      6.74702                population
11     2.88940     ocean_prox_NEAR OCEAN
9      2.53086         ocean_prox_ISLAND
7      2.05392             median_income
8      1.63289         ocean_prox_INLAND
13     1.56589         bedrooms_per_room
2      1.32363        housing_median_age
14     1.19886  population_per_household
10     1.06314       ocean_prox_NEAR BAY
12     1.00210       rooms_per_household

Removing the features with VIF>5 and p-values>0.05:

In [58]:
def get_vif(X_features):
    import statsmodels.api as sm
    from statsmodels.stats.outliers_influence import variance_inflation_factor
    l = [variance_inflation_factor(X_features.values, i) for i in range(X_features.shape[1])]
    return pd.Series(index=X_features.columns,data=l).sort_values(ascending=False)
            
def remove_by_pvalue(X_features,y_actual,pvalue=0.05):
    """Remove features with p-value more than 'pvalue'
    
    This function uses statsmodels.api.OLS model. Please add intercept to data externally.
    Input:
        X_features: Array or dataframe excluding predicted variable
        y_actual: Series or list of predicted variable
        pvalue: int or float
    
    Note:
        X_features is changed inplace
    """
    import statsmodels.api as sm
    for i in range(len(X_features.columns)):
        regressor_OLS=sm.OLS(endog=y_actual,exog=X_features).fit()
        s=regressor_OLS.pvalues.sort_values(ascending=False)
        if s.iloc[0]>pvalue:
            X_features.drop(s.index[0],axis=1,inplace=True)
            print('Removed: ',s.index[0],'P-value: ',s.iloc[0])
        
def remove_by_vif(X_features,vif=5):
    """Remove columns from X_features whose VIF is greater than supplied 'vif'
    Parameters:
        X_features:array or dataframe containing data excluding target variable
        vif: int or float of limiting value of VIF
    Note:
        This function changes X_features inplace
    """
    import statsmodels.api as sm
    from statsmodels.stats.outliers_influence import variance_inflation_factor
        
    for i in range(len(X_features.columns)):
        l = [variance_inflation_factor(X_features.values, i) for i in range(X_features.shape[1])]
        s=pd.Series(index=X_features.columns,data=l).sort_values(ascending=False)
        if s.iloc[0]>vif:
            X_features.drop(s.index[0],axis=1,inplace=True)
            print('Removed: ',s.index[0],', VIF: ',s.iloc[0])
        else:
            break

In [59]:
get_vif(df_Calidata2)

total_bedrooms              33.654665
households                  33.227681
latitude                    21.105923
longitude                   19.050705
total_rooms                 17.151574
population                   7.219240
median_income                3.297926
ocean_prox_NEAR OCEAN        2.950808
median_house_price           2.869236
ocean_prox_ISLAND            2.566745
ocean_prox_INLAND            1.643246
bedrooms_per_room            1.566232
housing_median_age           1.362382
population_per_household     1.199352
ocean_prox_NEAR BAY          1.063300
rooms_per_household          1.003294
dtype: float64

In [60]:
remove_by_vif(df_Calidata2)

Removed:  total_bedrooms , VIF:  33.65466536414311
Removed:  latitude , VIF:  21.09397600161885
Removed:  households , VIF:  15.190620598061336
Removed:  total_rooms , VIF:  5.650192291397495


Four columns are removed based on VIF constraints.

In [61]:
remove_by_pvalue(X_features,y_actual,pvalue=0.05)

Removed:  total_bedrooms P-value:  0.7255255841805073
Removed:  ocean_prox_NEAR BAY P-value:  0.07919154633467813
Removed:  total_rooms P-value:  0.05641978695219801


Three columns are removed based on p-values within 95% confidence level of interval.

Removing columns with high value of VIF:

In [62]:
df_Calidata2.head()

longitude  housing_median_age  population  median_income  \
0  -1.327835            0.982143   -0.974429       2.344766   
1  -1.322844           -0.607019    0.861439       2.332238   
2  -1.332827            1.856182   -0.820777       1.782699   
3  -1.337818            1.856182   -0.766028       0.932968   
4  -1.337818            1.856182   -0.759847      -0.012881   

   median_house_price  ocean_prox_INLAND  ocean_prox_ISLAND  \
0            2.129631           0.628559          -1.088291   
1            1.314156           0.327041          -0.936892   
2            1.258693           1.155620          -1.369108   
3            1.165100           0.156966          -0.465514   
4            1.172900           0.344711          -0.668831   

   ocean_prox_NEAR BAY  ocean_prox_NEAR OCEAN  rooms_per_household  \
0            -0.049597              -0.681889            -0.015566   
1            -0.092512              -0.681889            -0.015566   
2            -0.025843              -0.681889            -0.015566   
3            -0.050329              -0.681889            -0.015566   
4            -0.085616              -0.681889            -0.015566   

   bedrooms_per_room  population_per_household  
0           2.830742                 -0.384466  
1           2.830742                 -0.384466  
2           2.830742                 -0.384466  
3           2.830742                 -0.384466  
4           2.830742                 -0.384466

In [63]:
df_Calidata3 =df_Calidata2.drop(['ocean_prox_NEAR BAY'],axis=1)

In [64]:
df_Calidata3.head()

longitude  housing_median_age  population  median_income  \
0  -1.327835            0.982143   -0.974429       2.344766   
1  -1.322844           -0.607019    0.861439       2.332238   
2  -1.332827            1.856182   -0.820777       1.782699   
3  -1.337818            1.856182   -0.766028       0.932968   
4  -1.337818            1.856182   -0.759847      -0.012881   

   median_house_price  ocean_prox_INLAND  ocean_prox_ISLAND  \
0            2.129631           0.628559          -1.088291   
1            1.314156           0.327041          -0.936892   
2            1.258693           1.155620          -1.369108   
3            1.165100           0.156966          -0.465514   
4            1.172900           0.344711          -0.668831   

   ocean_prox_NEAR OCEAN  rooms_per_household  bedrooms_per_room  \
0              -0.681889            -0.015566           2.830742   
1              -0.681889            -0.015566           2.830742   
2              -0.681889            -0.015566           2.830742   
3              -0.681889            -0.015566           2.830742   
4              -0.681889            -0.015566           2.830742   

   population_per_household  
0                 -0.384466  
1                 -0.384466  
2                 -0.384466  
3                 -0.384466  
4                 -0.384466

In [65]:
df_Calidata3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 11 columns):
longitude                   20640 non-null float64
housing_median_age          20640 non-null float64
population                  20640 non-null float64
median_income               20640 non-null float64
median_house_price          20640 non-null float64
ocean_prox_INLAND           20640 non-null float64
ocean_prox_ISLAND           20640 non-null float64
ocean_prox_NEAR OCEAN       20640 non-null float64
rooms_per_household         20640 non-null float64
bedrooms_per_room           20640 non-null float64
population_per_household    20640 non-null float64
dtypes: float64(11)
memory usage: 1.7 MB


In [66]:
df_Calidata3.describe()

longitude  housing_median_age    population  median_income  \
count  2.064000e+04        2.064000e+04  2.064000e+04   2.064000e+04   
mean  -6.527810e-15        8.557001e-16 -6.465442e-17   3.734255e-16   
std    1.000024e+00        1.000024e+00  1.000024e+00   1.000024e+00   
min   -2.385992e+00       -2.196180e+00 -1.256123e+00  -1.774299e+00   
25%   -1.113209e+00       -8.453931e-01 -5.638089e-01  -6.881186e-01   
50%    5.389137e-01        2.864572e-02 -2.291318e-01  -1.767951e-01   
75%    7.784964e-01        6.643103e-01  2.644949e-01   4.593063e-01   
max    2.625280e+00        1.856182e+00  3.025033e+01   5.858286e+00   

       median_house_price  ocean_prox_INLAND  ocean_prox_ISLAND  \
count        2.064000e+04       2.064000e+04       2.064000e+04   
mean         8.950635e-16       2.704111e-16      -5.682712e-16   
std          1.000024e+00       1.000024e+00       1.000024e+00   
min         -1.662641e+00      -1.852319e+00      -3.105874e+00   
25%         -7.561633e-01      -3.994496e-01      -6.213544e-01   
50%         -2.353337e-01      -8.078489e-02      -1.613778e-01   
75%          5.014973e-01       2.519615e-01       4.428336e-01   
max          2.540411e+00       5.516324e+01       2.640304e+01   

       ocean_prox_NEAR OCEAN  rooms_per_household  bedrooms_per_room  \
count           2.064000e+04         2.064000e+04       2.064000e+04   
mean           -8.344316e-15         9.836823e-17       2.245947e-14   
std             1.000024e+00         1.000024e+00       1.000024e+00   
min            -6.818890e-01        -1.556621e-02      -3.532643e-01   
25%            -6.818890e-01        -1.556621e-02      -3.532643e-01   
50%            -6.818890e-01        -1.556621e-02      -3.532643e-01   
75%             1.466514e+00        -1.556621e-02      -3.532643e-01   
max             1.466514e+00         6.424173e+01       2.830742e+00   

       population_per_household  
count              2.064000e+04  
mean               4.781824e-14  
std                1.000024e+00  
min               -3.844665e-01  
25%               -3.844665e-01  
50%               -3.844665e-01  
75%               -3.844665e-01  
max                2.601007e+00

In [69]:
# Save dataframe as CSV
df_Calidata.to_csv('data_original.csv', columns = df_Calidata.columns)

In [70]:
# Save dataframe as CSV
df_Calidata3.to_csv('data_featured.csv', columns = df_Calidata3.columns)

In [ ]:
# Save dataframe as CSV
df_Calidata2.to_csv('data_wrangled.csv', columns = df_Calidata2.columns)